### PR TITLE
Storing Value Releation multi-selection as JSON

### PR DESCRIPTION
**Pull request by signedav**
If the type of the field is `QVariant::Map` - means a JSON source (from GPKG) - it reads and writes the values as JSON list. 

This works for GPKG JSON implementation #8707 and would work for Postgres JSON integration BUT Postgres JSON writing is not yet integrated.

### CODE REVIEW

File: src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp [hunk 1]
```diff
@@ -69,7 +69,22 @@ QVariant QgsValueRelationWidgetWrapper::value() const
...
+      foreach ( QString s, selection )
+      {
+        vl << s;
+      }
+      v = vl;
+    }
+    else
+    {
+      //store as hstore string
+      v = selection.join( QStringLiteral( "," ) ).prepend( '{' ).append( '}' );
```
Comment (m-kuhn) [inline src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp]:
```suggestion
       v = selection.join( QStringLiteral( ',' ) ).prepend( '{' ).append( '}' );
```

File: src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp [hunk 2]
```diff
@@ -69,7 +69,22 @@ QVariant QgsValueRelationWidgetWrapper::value() const
         }
       }
     }
-    v = selection.join( QStringLiteral( "," ) ).prepend( '{' ).append( '}' );
+
+    if ( layer()->fields().at( fieldIdx() ).type() == QVariant::Map )
+    {
+      QVariantList vl;
+      //store as QVariantList because it's json
+      foreach ( QString s, selection )
```
Comment (m-kuhn) [inline src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp]:
```suggestion
       for ( const QString &s, qgis::as_const( selection ) )
```

**Comment by m-kuhn** [comment]
Nice. Can we get a test for this in testqgsvaluerelationwidgetwrapper?

File: src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp [hunk 3]
```diff
@@ -69,7 +69,22 @@ QVariant QgsValueRelationWidgetWrapper::value() const
...
+      for ( const QString &s, qgis::as_const( selection ) )
+      {
+        vl << s;
+      }
+      v = vl;
+    }
+    else
+    {
+      //store as hstore string
+      v = selection.join( QStringLiteral( ',' ) ).prepend( '{' ).append( '}' );
```
Comment (elpaso) [inline src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp]: More a question than a comment: is `QStringLiteral` really useful here?

Comment (m-kuhn) [reply src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp]: Not actually, it's a leftover of trying to fix things online https://github.com/qgis/QGIS/pull/8721/commits/eb9310fec88c851e8a147e8b6ef0480756e51342 ;P

Comment (m-kuhn) [inline src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp]:
```suggestion
       v = selection.join( ',' ).prepend( '{' ).append( '}' );
```

File: tests/src/gui/testqgsvaluerelationwidgetwrapper.cpp [hunk 4]
```diff
@@ -368,5 +373,179 @@ void TestQgsValueRelationWidgetWrapper::testZeroIndexInRelatedTable()
...
+  QCOMPARE( w_favoriteauthors.mTableWidget->item( 4, 0 )->text(), QStringLiteral( "Douglas Adams" ) );
+  QCOMPARE( w_favoriteauthors.mTableWidget->item( 4, 0 )->checkState(), Qt::Checked );
+  QCOMPARE( w_favoriteauthors.mTableWidget->item( 5, 0 )->text(), QStringLiteral( "Ken Follett" ) );
+  QCOMPARE( w_favoriteauthors.mTableWidget->item( 5, 0 )->checkState(), Qt::Checked );
+}
+
+
+void TestQgsValueRelationWidgetWrapper::testWithJsonInGPKG()
+{
+  /* prepared - functionality needs to be integrated first
```
Comment (m-kuhn) [inline tests/src/gui/testqgsvaluerelationwidgetwrapper.cpp]: Replace comment with `#if GDAL_VERSION...`

File: tests/src/gui/testqgsvaluerelationwidgetwrapper.cpp [hunk 5]
```diff
@@ -368,5 +374,246 @@ void TestQgsValueRelationWidgetWrapper::testZeroIndexInRelatedTable()
...
+  QCOMPARE( w_favoriteauthors_b.mTableWidget->item( 4, 0 )->text(), QStringLiteral( "Douglas Adams" ) );
+  QCOMPARE( w_favoriteauthors_b.mTableWidget->item( 4, 0 )->checkState(), Qt::Checked );
+  QCOMPARE( w_favoriteauthors_b.mTableWidget->item( 5, 0 )->text(), QStringLiteral( "Ken Follett" ) );
+  QCOMPARE( w_favoriteauthors_b.mTableWidget->item( 5, 0 )->checkState(), Qt::Checked );
+}
+
+
+void TestQgsValueRelationWidgetWrapper::testWithJsonInGPKG()
+{
+#if GDAL_VERSION_NUM >= GDAL_COMPUTE_VERSION(2,4,0)
```
Comment (m-kuhn) [inline tests/src/gui/testqgsvaluerelationwidgetwrapper.cpp]: @signedav did you run those tests with GDAL 2.4 locally?

Comment (signedav) [reply tests/src/gui/testqgsvaluerelationwidgetwrapper.cpp]: Yes.

**Comment by signedav** [comment]
:tada:

**Pull request merged**